tokio/src/sync/broadcast.rs

```diff
@@ -544,6 +544,8 @@ impl<T> Sender<T> {
         let mut tail = self.shared.tail.lock();
 
         if tail.rx_cnt == 0 {
+            // Go back to drop values earlier and potentially consume less memory
+            tail.pos = 0;
             return Err(SendError(value));
         }
```

Comment:
Are you targeting the scenario where the channel is reopened later, or when messages have already been sent before it was closed?

Reply:
The scenario where the channel might reopen later, because all receivers were dropped (or never existed) and queued messages are unreachable.

And by "reopening" I mean calling `.subscribe()` again.

Reply:
It seems like it would make more sense to do this in `new_receiver` instead? Then it also applies when all receivers are dropped, then a new one is created, without an attempt to send between those two events.

Reply:
You're right, makes more sense to do this when new receivers are created.

--- 

Idk, this PR is a small patch to a corner case.

I was thinking, is there another good way of using less memory? Maybe something like:

```rust
/// slots in the channel.
buffer: VecDeque<RwLock<Slot<T>>,
```

So we could have an unbounded broadcast channel, and, most importantly in my opinion: a bounded broadcast channel that **does not** consume all the memory available, even if receivers are keeping up.

Currently, if I have 1024 channels with a buffer size of 4kb, it will all consume 4MB of RAM before channels are even used, that's because buffers are allocated and initialized, which forces the OS to give a real and free piece of available RAM for that virtual address, instead of lazily providing it.

By using VecDeque we could get away with amortized growth when growing is necessary (just like Vec), and empty VecDeques take 0 space in the heap, any opinions on this? Do you have a good alternate design in mind or is it a bad idea?

Reply:
There are many different ways of implementing channels with different tradeoffs. What you're suggesting requires extra synchronization around the `buffer`, because you need mutable access to resize it. I'm reluctant to add that extra synchronization.